flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/AlternatingCheckpointBarrierHandlerTest.java

```diff
@@ -0,0 +1,212 @@
+		}
+	}
+
+	private static AlternatingCheckpointBarrierHandler barrierHandler(SingleInputGate inputGate, AbstractInvokable target) {
+		String taskName = "test";
+		InputGate[] channelIndexToInputGate = new InputGate[inputGate.getNumberOfInputChannels()];
+		Arrays.fill(channelIndexToInputGate, inputGate);
+		return new AlternatingCheckpointBarrierHandler(
+				new CheckpointBarrierAligner(taskName, channelIndexToInputGate, singletonMap(inputGate, 0), target),
+				new CheckpointBarrierUnaligner(new int[]{inputGate.getNumberOfInputChannels()}, ChannelStateWriter.NO_OP, taskName, target),
+				target);
+	}
+
+	private Buffer barrier(long id, CheckpointType checkpointType) throws IOException {
+		return toBuffer(new CheckpointBarrier(
+				id,
+				System.currentTimeMillis(),
+				new CheckpointOptions(checkpointType, CheckpointStorageLocationReference.getDefault(), true, true)));
+	}
+
+	private static class TestInvokable extends AbstractInvokable {
+		private List<Long> triggeredCheckpoints = new ArrayList<>();
+
+		TestInvokable() {
+			super(new DummyEnvironment());
+		}
+
+		@Override
+		public void invoke() {
+		}
+
+		@Override
+		public <E extends Exception> void executeInTaskThread(ThrowingRunnable<E> runnable, String descriptionFormat, Object... descriptionArgs) throws E {
+			runnable.run();
+		}
+
+		@Override
+		public void triggerCheckpointOnBarrier(CheckpointMetaData checkpointMetaData, CheckpointOptions checkpointOptions, CheckpointMetrics checkpointMetrics) {
+			triggeredCheckpoints.add(checkpointMetaData.getCheckpointId());
+		}
+	}
+
+	private static class TestInputChannel extends InputChannel {
+		private final Deque<Buffer> buffers = new LinkedList<>();
+		private boolean wasResumed;
+
+		TestInputChannel(SingleInputGate inputGate, int channelIndex) {
+			super(inputGate, channelIndex, new ResultPartitionID(), 0, 10, null, null);
+		}
+
+		public void receive(Buffer buffer) {
+			buffers.add(buffer);
+			notifyChannelNonEmpty();
+		}
+
+		@Override
+		public void requestSubpartition(int subpartitionIndex) {
+		}
+
+		@Override
+		protected Optional<BufferAndAvailability> getNextBuffer() {
+			return buffers.isEmpty() ? Optional.empty() : Optional.of(new InputChannel.BufferAndAvailability(buffers.poll(), buffers.isEmpty(), buffers.size()));
+		}
+
+		@Override
+		protected void sendTaskEvent(TaskEvent event) {
+		}
+
+		@Override
+		protected boolean isReleased() {
+			return false;
+		}
+
+		@Override
+		protected void releaseAllResources() {
+		}
+
+		@Override
+		public void resumeConsumption() {
+			wasResumed = true;
+		}
+	}
+
+	private static CheckpointedInputGate buildGate(TestInvokable target, int numChannels) {
+		SingleInputGate gate = new SingleInputGateBuilder().setNumberOfChannels(numChannels).build();
+		TestInputChannel[] channels = new TestInputChannel[numChannels];
+		for (int i = 0; i < numChannels; i++) {
+			channels[i] = new TestInputChannel(gate, i);
+		}
+		gate.setInputChannels(channels);
```

Comment:
nit: could be a oneliner with (Int)Stream.

Reply:
In this case it didn't improve readability.